File: desktop-mend/src/test/java/commands/TestBase.java
package commands;

import co.samco.mend4.core.OSDao;
import co.samco.mend4.core.Settings;
import co.samco.mend4.core.crypto.CryptoProvider;
import co.samco.mend4.desktop.commands.Command;
import co.samco.mend4.desktop.core.I18N;
import co.samco.mend4.desktop.exception.SettingRequiredException;
import co.samco.mend4.desktop.helper.*;
import co.samco.mend4.desktop.output.PrintStreamProvider;
import org.junit.Assert;
import org.junit.Before;

import java.io.IOException;
import java.io.PrintStream;
import java.util.ArrayList;

import static org.mockito.Mockito.*;

public class TestBase {
    protected CryptoHelper cryptoHelper;
    protected CryptoProvider cryptoProvider;
    protected FileResolveHelper fileResolveHelper;
    protected InputHelper inputHelper;
    protected KeyHelper keyHelper;
    protected MergeHelper mergeHelper;
    protected SettingsHelper settingsHelper;
    protected ShredHelper shredHelper;
    protected VersionHelper versionHelper;
    protected OSDao osDao;
    protected I18N strings;
    protected PrintStreamProvider log;
    protected PrintStream out;
    protected PrintStream err;
    protected Settings settings;

    @Before
    public void setup() {
        strings = new I18N("en", "UK");
        err = mock(PrintStream.class);
        out = mock(PrintStream.class);
        log = mock(PrintStreamProvider.class);
        when(log.err()).thenReturn(err);
        when(log.out()).thenReturn(out);
        osDao = mock(OSDao.class);
        cryptoHelper = mock(CryptoHelper.class);
        cryptoProvider = mock(CryptoProvider.class);
        fileResolveHelper = mock(FileResolveHelper.class);
        inputHelper = mock(InputHelper.class);
        keyHelper = mock(KeyHelper.class);
        mergeHelper = mock(MergeHelper.class);
        settingsHelper = mock(SettingsHelper.class);
        shredHelper = mock(ShredHelper.class);
        versionHelper = mock(VersionHelper.class);
        settings = mock(Settings.class);
    }

    protected void testCommandWithSettingsDependencies(Command command) throws IOException, SettingRequiredException {
        String message = "message";
        doThrow(new SettingRequiredException(message)).when(settingsHelper).assertRequiredSettingsExist(any(), any());
        command.executeCommand(new ArrayList<>());
        verify(err).println(eq(message));
        Assert.assertNotEquals(0, command.getExecutionResult());
    }
}

File: desktop-mend/src/test/java/commands/TestBase.kt
package commands

import co.samco.mend4.core.crypto.CryptoProvider
import co.samco.mend4.desktop.commands.Command
import co.samco.mend4.desktop.core.I18N
import co.samco.mend4.desktop.core.I18NImpl
import co.samco.mend4.desktop.dao.OSDao
import co.samco.mend4.desktop.dao.SettingsDao
import co.samco.mend4.desktop.exception.SettingRequiredException
import co.samco.mend4.desktop.helper.CryptoHelper
import co.samco.mend4.desktop.helper.FileResolveHelper
import co.samco.mend4.desktop.helper.InputHelper
import co.samco.mend4.desktop.helper.KeyHelper
import co.samco.mend4.desktop.helper.MergeHelper
import co.samco.mend4.desktop.helper.SettingsHelper
import co.samco.mend4.desktop.helper.ShredHelper
import co.samco.mend4.desktop.helper.VersionHelper
import co.samco.mend4.desktop.output.PrintStreamProvider
import org.junit.Assert
import org.junit.Before
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.doThrow
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import java.io.IOException
import java.io.PrintStream

open class TestBase {

    protected lateinit var cryptoHelper: CryptoHelper

    protected lateinit var cryptoProvider: CryptoProvider

    protected lateinit var fileResolveHelper: FileResolveHelper

    protected lateinit var inputHelper: InputHelper

    protected lateinit var keyHelper: KeyHelper

    protected lateinit var mergeHelper: MergeHelper

    protected lateinit var settingsHelper: SettingsHelper

    protected lateinit var shredHelper: ShredHelper

    protected lateinit var versionHelper: VersionHelper

    protected lateinit var strings: I18N

    protected lateinit var log: PrintStreamProvider

    protected lateinit var out: PrintStream

    protected lateinit var err: PrintStream

    protected lateinit var settings: SettingsDao

    protected lateinit var osDao: OSDao

    @Before
    open fun setup() {
        err = mock()
        out = mock()
        log = mock()
        whenever(log.err()).thenReturn(err)
        whenever(log.out()).thenReturn(out)
        cryptoProvider = mock()
        settings = mock()
        osDao = mock()
        settingsHelper = mock()
        versionHelper = mock()
        strings = I18NImpl("en", "UK")
        fileResolveHelper = mock()
        inputHelper = mock()
        keyHelper = mock()
        mergeHelper = mock()
        cryptoHelper = mock()
        shredHelper = mock()
    }

    @Throws(IOException::class, SettingRequiredException::class)
    protected fun testCommandWithNoSettingsDependencies(command: Command) {
        val message = "message"
        doThrow(SettingRequiredException(message))
            .whenever(settingsHelper)
            .assertRequiredSettingsExist(any(), any())

        command.executeCommand(ArrayList())
        verify(err).println(eq(message))
        Assert.assertNotEquals(0, command.executionResult.toLong())
    }
}